packages/grafana-runtime/src/utils/queryResponse.ts

```diff
@@ -127,6 +127,36 @@ export function toDataQueryResponse(
   return rsp;
 }
 
+/**
+ * Data sources using api/ds/query to test data sources can use this function to
+ * handle errors and convert them to TestingStatus object.
+ *
+ * If possible, this should be avoided in favor of implementing /health endpoint
+ * and testing data source with DataSourceWithBackend.testDataSource()
+ *
+ * @returns {TestingStatus}
+ */
+export function toTestingStatus(err: FetchError): any {
+  const queryResponse = toDataQueryResponse(err);
+  // POST api/ds/query errors returned as { message: string, error: string } objects
+  if (queryResponse.error?.data?.message && queryResponse.error?.data?.error) {
```

Comment:
I think this logic is wrong since error is only included when [app_mode](https://grafana.com/docs/grafana/latest/administration/configuration/#app_mode) is development. But I might be wrong since not sure if `toDataQueryResponse` somehow handles this case and always populates the `data.error`?

```suggestion
  if (queryResponse.error?.data?.message) {
```